packages/cmr-client/src/EarthdataLogin.ts

```diff
@@ -1,202 +1,155 @@
+        username: username,
+        password: password,
+      });
+  } catch (error) {
+    if (error instanceof got.HTTPError) throw parseHttpError(error, 'retrieve');
+    throw parseCaughtError(error);
   }
+  const tokens = GetTokenResponseBody.parse(rawResponse.body);
+  const isTokenExpired = (token: Token) => new Date(token.expiration_date) < new Date();
+  const unExpiredTokens = tokens.filter((token: any) => !isTokenExpired(token));
+  return unExpiredTokens.length > 0 ? sortBy(unExpiredTokens, ['expiration_date'])[0].access_token : undefined;
+};
 
-  /**
-   * The method for getting the token from the Earthdata Login endpoint. Sends a GET request
-   * with the users' base64 encoded username and password as a header for authorization. If the
-   * users' credentials are accepted the first unexpired token is retrieved, if one exists, and
-   * returned on their behalf, if not, an error is thrown. If the user does not have a token
-   * in Earthdata Login then undefined is returned to indicate to token needs to be created.
-   *
-   * @returns {Promise.<string>} the token
-   */
-  async retrieveEDLToken(): Promise<string> {
-    const buff = Buffer.from(`${this.username + ':' + this.password}`).toString('base64');
-    const url = this.getEDLurl();
-    // response: get a token from the Earthdata login endpoint using credentials if exists
-    let response: EarthdataGetTokenResponse;
-    try {
-      response = await got.get(`${url}/api/users/tokens`,
-        {
-          headers: {
-            Authorization: `Basic ${buff}`,
-          },
-        }).json();
-    } catch (error) {
-      this.handleHttpError(error);
-    }
-    const currDate = new Date();
-
-    for (let i = 0; i < Object.keys(response).length; i += 1) {
-      const responseDate = new Date(response[i].expiration_date);
-      if (currDate < responseDate) {
-        return response[i].access_token;
-      }
-    }
-    return undefined;
+/**
+  * The method for creating Earthdata Login token. This method sends a POST request
+  * to the Earthdata Login endpoint URL in order to create a token for the user. The users'
+  * username and password are sent as base64 encoded credentials as a header for
+  * authorization. If the users' credentials are accepted a token is created on their
+  * behalf and returned, if not, an error is thrown.
+  *
+  * @returns {string} the token
+*/
+export const createEDLToken = async (username: string, password: string, edlEnv: string):
+Promise<string> => {
+  let rawResponse: any;
+  try {
+    rawResponse = await got.post(`${getEdlUrl(edlEnv)}/api/users/token`,
+      {
+        responseType: 'json',
+        username: username,
+        password: password,
+      });
+  } catch (error) {
+    if (error instanceof got.HTTPError) throw parseHttpError(error, 'create');
+    throw parseCaughtError(error);
   }
+  const response = PostTokenResponseBody.parse(rawResponse.body);
+  return response[0].access_token;
```

Comment:
Without a check for the number of elements in `response`, `response[0]` is potentially `undefined`.

Reply:
If that's the case, then `PostTokenResponseBody` is never going to match. That schema is looking for an array, but the docs you link to here show an object being returned.

It also looks like the tests are using the incorrect value. They are sending an array, but the linked docs show an object being returned.

Reply:
If it's a string then it's likely JSON that needs to be parsed using `JSON.parse`. That being said, the `got` client should be parsing that JSON response if you've specified `responseType: 'json'`.